### PR TITLE
[WIP] Add fixed subtitle to home page

### PR DESCRIPTION
**Pull request by lmorchard**
- New Visibility component that changes class name based on visibility

- Move HomePage components into bundle

- Wrap experiments list on home page in Visibility component, styles to
  make the section subtitle appear fixed when the section is out of view

Issue #3001

### CODE REVIEW

**Comment by lmorchard** [comment]
Still might need some work on a couple of things:
- [ ] Couldn't find an exact spec for this, so I just guessed on styles
- [ ] Didn't yet change the header content to "Or try our other experiments" - but will need l10n when changed

File: frontend/src/app/components/Visibility/index.js
Comment (lmorchard) [inline frontend/src/app/components/Visibility/index.js:16]: Also "Visibility" might be too general a name for this utility component, since so far it only checks to see if the element is vertically higher than current scroll position. Doesn't do the inverse - which could be handy for simplifying the sticky header on experiment pages. Figured that could be for a future enhancement though

**Comment by lmorchard** [comment]
Also looks like this will collide with PR #3114 thanks to the HomePage components move. But the changes here to those components are small, so we could put this PR off until that one's merged

**Comment by MozillaTestPilotStorybook** [comment]
# Storybook Deployment
Pull Request: http://testpilot-storybook.dev.mozaws.net/pull/3143/index.html
Commit: http://testpilot-storybook.dev.mozaws.net/fa3eabbf8202f827184034b52ed549107511e4ad/

**Comment by lmorchard** [comment]
[Here's a direct link to the feature in the last storybook build](http://testpilot-storybook.dev.mozaws.net/fa3eabbf8202f827184034b52ed549107511e4ad/?selectedKind=HomePage&selectedStory=Fixed%20off-screen%20subtitle&full=0&down=1&left=1&panelRight=0&downPanel=storybook%2Factions%2Factions-panel)

**Comment by SoftVision-PaulOiegas** [comment]
Just an observation. After looking on storybook animation for the fixed subtitle I've noticed that the attach and detach part is not really smooth when slow scrolling. 
See this recording https://goo.gl/jAF5DY.

**Comment by lmorchard** [comment]
Yeah, there is no animation or transition at all right now. Didn't want to put too much into that without an idea of how the styles should work overall

**Comment by fzzzy** [comment]
I did a very similar thing for pageshot -- what I did was set visibility: hidden on the element which is in the scrolling part (so that the page would continue to lay out as if that element was there), and then wait until that element is exactly where the non-scrolling element is. Then set visibility: visible on the scrolling element and display: none on the fixed element.

File: frontend/src/app/containers/HomePage/index.scss
Comment (fzzzy) [inline frontend/src/app/containers/HomePage/index.scss:13]: This background being transparent doesn't really work; it can clash with whatever happens to be in the page.
screen shot 2017-12-18 at 11 26 32 am

**Review by fzzzy — CHANGES_REQUESTED** [review]
At the very least I think the background of the banner needs to be opaque. If, once the background is opaque, the fixed element could be made to not disappear until the scrolling element is exactly in the same place, that would be good, but I tried to make that change to the logic and it did not appear to be straightforward to accomplish.

**Comment by lmorchard** [comment]
Looks like this animation spec is related to this issue, but has a curve to the banner I haven't seen before. Thinking that's the UX spec I'm missing https://github.com/mozilla/testpilot/issues/3039

File: frontend/src/app/containers/HomePage/index.scss
Comment (lmorchard) [reply frontend/src/app/containers/HomePage/index.scss:13]: Yeah, like I said in a previous comment, I didn't put much work into the styles because I don't have a UX spec for it. I don't think the styles I have are good at all. This is mainly just to get it "functional"

**Comment by lmorchard** [comment]
Just added a "WIP" to the title - let's not merge this yet, it needs some UX work

File: frontend/src/app/components/Visibility/tests.js
Comment (pdehaan) [inline frontend/src/app/components/Visibility/tests.js:1]: Instead of specifying the methods individually, we could just set the `mocha` ESLint environment: `/* eslint-env mocha */`

Or, possibly look into using ESLint config `overrides` with a glob to automatically set the `mocha` env for `frontend/src/**/tests.js` files:
https://eslint.org/docs/user-guide/configuring#configuration-based-on-glob-patterns

**Comment by johngruen** [comment]
Okay, finally looking at this

File: frontend/src/app/components/Visibility/tests.js
Comment (johngruen) [reply frontend/src/app/components/Visibility/tests.js:1]: Yeah, that's a good idea....could you file an issue to track?

Comment (johngruen) [reply frontend/src/app/components/Visibility/tests.js:1]: b/c we do this in all the places

**Comment by lmorchard** [comment]
Going to close this for now, and resubmit with updated UX

**Pull request closed without merging**
